cache/merge.go

```diff
@@ -46,8 +55,13 @@ func BuildTop(cacheReader CacheReader) (*Reader, []byte, error) {
 		}
 	}
 
-	// Create a new subtree with the cache highest layer as its leaves.
-	subtreeWriter := NewWriter(MinHeightPolicy(0), MakeSliceReadWriterFactory())
+	// Create an adjusted caching policy for the new subtree.
+	newCachingPolicy := func(layerHeight uint) bool {
+		return cacheReader.GetCachingPolicy()(maxHeight + layerHeight)
+	}
+
+	// Create a subtree with the cache highest layer as its leaves.
+	subtreeWriter := NewWriter(newCachingPolicy, cacheReader.GetLayerFactory())
```

Comment:
The layer factory should also be wrapped like the caching policy. It accepts a height and may use that, for example, to generate a filename or to generate different types of layers per height (both use-cases are employed in PoST).